### ISSUE TITLE
Replace script packing system

### ISSUE THREAD

**Issue by SaintWish** [MSRevive/MasterSwordRebirth#248]
At some point it'd be nice to replace the script packing system with a Pak file sort of system, like here (https://simoncoenen.com/blog/programming/PakFiles). The main benefit is we'd be able to compress it. This should probably be done after Angelscript is put in.

We need to make sure we only load the header part of the Pak and then load script files when needed.

**Comment by tschumann** [comment]
Is there any reason that it needs to be in a .pak file besides compression? Could the files be loose?

Is the reason they're packed in the .dll now to prevent tampering?

**Comment by SaintWish** [comment]
The reason we want to pack the script files up using some kind of common format like PAK or just a flat out ZIP file is so we can perform a CRC check on the pack against the server to make sure the client didn't temper with their scripts.

The reason they're packed in a DLL is from MSC legacy code and IIRC they did it to prevent tempering and they did encrypt them (which we've removed).

**Comment by BenjaBobs** [comment]
Iirc there was also originally an element of closed source for the purpose making it harder to exploit things and discover bugs, but I don't know how relevant or effective it was.

**Comment by tschumann** [comment]
Okay makes sense - should be doable. The engine code has some CRC functions but I don't think they're exposed on the client. https://github.com/ValveSoftware/halflife/blob/master/game_shared/simple_checksum.h looks quite primitive but might be a quick and easy option.

Maybe load the .pak on the server, compute its checksum, then transmit that to clients on game join and kick if it doesn't match.

**Comment by SaintWish** [comment]
Yeah currently we check the CRC against what the FN server has stored using Source's CRC32 and we also use this [function](https://github.com/MSRevive/MasterSwordRebirth/blob/4e2048da69a38f4dc85cbdfac75fbe412c59e754/src/game/server/svglobals.cpp#L163) to make sure client's matches up with the server's.

But yeah it's planned to implement this along side Angelscript or immediately after Angelscript or that's the hope anyways lol.

**Comment by tschumann** [comment]
Ah yeah, should be possible to do that with a .pak file then - I knew the
engine supported CRC checks but I didn't realise it could be used for
arbitrary files.

**Comment by tschumann** [comment]
https://github.com/MSRevive/MasterSwordRebirth/pull/269 is untested but an attempt at replacing it with pak0.pak - I still need to test it (and also track down a command line .pak file creator).

Thoughts? It can be expanded on later but this should simplify the script loading and also speed it up a bit (looks like there was some double-handling of memory on load).

**Comment by SaintWish** [comment]
If you're unable to track down a CLI pak program I can probably create one at some point. But cheers for doing this; this will help a lot with switching to STL types and such.

**Comment by tschumann** [comment]
Actually thinking about it some more, I might restore scriptpack and just get it to create the .pak file - it can then be expanded later with compression or whatever.

**Comment by SaintWish** [comment]
Sounds good to me
